2023-07-28-on-demand-builders.md

```diff
@@ -0,0 +1,125 @@
+
+The problem
+-----------
+
+The thing is that, as of now, every builder in Copr is started by Resalloc "in
+advance", or "preallocated".  The builders are preallocated even though they are
+not _immediately_ used, and stay idle till someone takes them (if there's no
+demand for days, they stay allocated).  When taken, system allocates additional
+resources (so we get more and more machines if there's actual demand for them).
+
+This system works well for normal builders, because they are typically taken
+very quickly (the more tasks in queue the faster they are taken; the smaller the
+"prealloc" quota number for resalloc pool is the faster they are taken).  And
+even if they are not taken for a while, we can actually afford keeping a few of
+them allocated because the instances are not overly expensive.
+
+There's though a demand to have a more (maybe ridiculously) powerful instances;
+having just one of them allocated but unused could be quite expensive matter.
+Even if we used for a low-demanding task.
+
+
+The proposal overview
+---------------------
+
+Let's have on-demand resources/pools in Resalloc.  In such pools there would be
+**no preallocated resource** by default, as long as there's no demand for them.
+Client will have a possibility to ask for them via special "on demand" tagged
+tickets.
+
+We want to be able to
+
+    - allow-list for supported architectures (because for some
+      architectures like s390x it might be quite hard to get "extra_powerful"
+      machines).
+
+    - allowlist the sets of users that can use the tags, preferably
+      through OIDC/FAS/LDAP groups -> we might want to have some acknowledge
+      from those users that the feature won't be overused or misused
+
+Resalloc Implementation
+-----------------------
+
+When a ticket with such a tag is taken from Resalloc, the resource manager would
+register such a ticket and then start allocating the expensive resource.
+
+Multiple pools can provide "on demand" tagged resources.
+
+If the ticket is closed early (user/copr changed the mind), the machine would be
+deallocated to save the money.
+
+The client API is unchanged, just using this as normally:
+
+    $ resalloc ticket --tag arch_x86_64 --tag extra_powerful
+    ID
+    $ ticket close ID
+
+
+Initial draft proposal: https://github.com/praiskup/resalloc/pull/118
+
+
+Copr implementation
+-------------------
+
+Let's have a new `copr build --builder-tag TAG` option (multiple tags may be
+specified).
+
+- non-existing tags should throw an error
+- permission error might be raised, if user isn't allowed to use it
+- warning can be thrown if TAG is not matching all the selected chroots
+  (architectures)
+
+New table **BuilderTag** should be created in database
+
+    - **ID** - integer
+    - **Name** - non-null, unique
+    - **groups_allowed** - list of (e.g. FAS) groups that can use this
+    - **achr_allowlist** - list of architectures this tag is supported for (inlcuding noarch)
```

Comment:
I am a bit confused by this. Is this table going to have just one row?

| ID | name | groups_allowed | arch_allowlist |
|--- | --- | --- | --- | 
| 1 | beefy | @group1, @group2, @group3 | x86_64, aarch64 |

Two more thoughts:

1. Having also `users_allowed` would probably be a good idea
2. I am not sure if we need allow list or rather deny list of bad users.

Reply:
> just one row?

Yes, that's what I meant.

> users_allowed

This has the potential to complicate our maintenance with per-user settings;  I'm not against it, but I'd maybe wait with this addition till we are sure it is needed.

Reply:
> deny list of bad users.

That was one of the initial requirements from @xsuchy, and I can't decide myself now.  I actually think that with the beefy builders, we'll have to be much more careful about allocation and users should know that, maybe rather explicitly (actual builders are not slow, sure, but the overall user feeling must be that they are for free, because most of them hardly touch any limit).